[Intelligence Effects] Smart Replies animation flickers / doesn't animate smoothly

rr-codes opened the pull request:
#### 3b6a05692af693c269419e4486ddd1ac1190d5a6
<pre>
[Intelligence Effects] Smart Replies animation flickers / doesn&apos;t animate smoothly
<a href="https://bugs.webkit.org/show_bug.cgi?id=286231">https://bugs.webkit.org/show_bug.cgi?id=286231</a>
<a href="https://rdar.apple.com/143206029">rdar://143206029</a>

Reviewed by Aditya Keerthi.

Refactor the Smart Replies implementation to use the new intelligence effects system, which fixes the existing
flickering and broken animations in the current implementation.

* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::DocumentMarkerController::addMarker):
(WebCore::DocumentMarkerController::addTransparentContentMarker):
(WebCore::DocumentMarkerController::applyToCollapsedRangeMarker):
(WebCore::DocumentMarkerController::forEach&lt;DocumentMarkerController::IterationDirection::Forwards&gt;):
(WebCore::DocumentMarkerController::forEach&lt;DocumentMarkerController::IterationDirection::Backwards&gt;):
(WebCore::addMarker):
* Source/WebCore/dom/DocumentMarkerController.h:
(WebCore::addMarker):
* Source/WebCore/page/writing-tools/WritingToolsController.h:
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::smartReplySessionDidReceiveTextWithReplacementRange):
(WebCore::WritingToolsController::compositionSessionDidReceiveTextWithReplacementRange):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView didBeginWritingToolsSession:contexts:]):
(-[WKWebView didEndWritingToolsSession:accepted:]):
(-[WKWebView compositionSession:didReceiveText:replacementRange:inContext:finished:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::compositionSessionDidReceiveTextWithReplacementRange):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/WebKitSwift.h:
* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceSmartReplyTextEffectCoordinator.h: Copied from Source/WebKit/WebKitSwift/WebKitSwift.h.
* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceSmartReplyTextEffectCoordinator.swift: Added.
(WKIntelligenceSmartReplyTextEffectCoordinator.onFlushCompletion):
(WKIntelligenceSmartReplyTextEffectCoordinator.hasActiveEffects):
(WKIntelligenceSmartReplyTextEffectCoordinator.startAnimation(for:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.requestReplacement(withProcessedRange:finished:characterDelta:operation:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.flushReplacements):
(WKIntelligenceSmartReplyTextEffectCoordinator.restoreSelection(_:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.hideEffects):
(WKIntelligenceSmartReplyTextEffectCoordinator.showEffects):
(WKIntelligenceSmartReplyTextEffectCoordinator.removeActiveEffects):
(WKIntelligenceSmartReplyTextEffectCoordinator.startReplacementAnimation(withProcessedRange:characterDelta:operation:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.reset):
(WKIntelligenceSmartReplyTextEffectCoordinator.updateTextChunkVisibility(_:visible:force:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.textPreview(for:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.updateTextChunkVisibility(_:visible:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.performReplacementAndGeneratePreview(for:effect:remainder:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.replacementEffectWillBegin(_:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.replacementEffectDidComplete(_:)):
(async(_:)):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::compositionSessionDidReceiveTextWithReplacementRange):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/289354@main">https://commits.webkit.org/289354@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f49f762b85b8263ac0a6f4b83b2b2e063635706b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86652 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6159 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40984 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91496 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/37384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88701 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6427 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14217 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/67014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/37384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89655 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/4884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/78461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/47337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/4683 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/32792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36502 "Failed to checkout and rebase branch from PR 39269") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/75134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/33677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93370 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13805 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/10010 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/75809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14006 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/74283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/74996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/19296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/17701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6590 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13465 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13828 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/19088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13566 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17010 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15350 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/dom/DocumentMarkerController.cpp:
```diff
@@ -510,6 +529,28 @@ Vector<WeakPtr<RenderedDocumentMarker>> DocumentMarkerController::markersFor(Nod
+// Finds any document markers of the specified types that contain a collapsed range matching `range`,
+// and applies `function` to them.
+//
+// If `range` is not collapsed, this function does not do anything.
```
This is more of a 'what' comment than a why comment. I would either remove it entirely, or adjust it to explain why we ignore collapsed ranges (if the reason isn't immediately apparent).

whsieh commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -1054,6 +1140,13 @@ static bool appliedCommandIsWritingToolsCommand(const Vector<Ref<WritingToolsCom
+    if (state->session.compositionType == WritingTools::SessionCompositionType::SmartReply) {
+        // The restart side effects happen in the beginning of `didReceiveText` instead.
+        // This is so that the replacement queue can operate correctly since it is async.
+
```
```suggestion
```

whsieh commented on Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceSmartReplyTextEffectCoordinator.h:
```diff
@@ -0,0 +1,43 @@
+
+#import <Foundation/Foundation.h>
+
+#if !TARGET_OS_WATCH && !TARGET_OS_TV && __has_include(<WritingTools/WritingTools.h>)
```
Can we use `PLATFORM(WATCHOS)` and `PLATFORM(APPLETV)` here, or is this an SPI header (i.e. usable by WebKit clients) where we can't include WTF headers?

rr-codes replied:
I don't think so, since this is all within WebKitSwift (I also checked and there doesn't seem to be any uses of PLATFORM inside WKS). This is also consistent with other header files in WKS.

pxlcoder commented on Source/WebCore/dom/DocumentMarkerController.cpp:
```diff
@@ -510,6 +529,26 @@ Vector<WeakPtr<RenderedDocumentMarker>> DocumentMarkerController::markersFor(Nod
 
+// Finds any document markers of the specified types that contain a collapsed range matching `range`,
+// and applies `function` to them.
+void DocumentMarkerController::applyToCollapsedRangeMarker(const SimpleRange& range, OptionSet<DocumentMarkerType> types, Function<bool(Node&, RenderedDocumentMarker&)>&& function)
```
Should this early return if `types` doesn't contain `DocumentMarkerType::TransparentContent`?

rr-codes replied:
I don't think it's needed, since you can't insert a collapsed DocumentMarker that isn't of that type in the first place, so this function will never do anything for other types

pxlcoder commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -508,6 +508,73 @@ static bool isZeroToOneCompositionType(WritingTools::Session::CompositionType ty
+
+    Ref currentCommand = state->reappliedCommands.takeLast();
+
+    // The prior replacement command must be undone in such a way as to not have it be added to the undo stack
```
```suggestion
    // The prior replacement command must be undone in such a way as to not have it be added to the undo stack.
```

rr-codes committed:
[Intelligence Effects] Smart Replies animation flickers / doesn't animate smoothly
https://bugs.webkit.org/show_bug.cgi?id=286231
rdar://143206029

Reviewed by Aditya Keerthi.

Refactor the Smart Replies implementation to use the new intelligence effects system, which fixes the existing
flickering and broken animations in the current implementation.

* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::DocumentMarkerController::addMarker):
(WebCore::DocumentMarkerController::addTransparentContentMarker):
(WebCore::DocumentMarkerController::applyToCollapsedRangeMarker):
(WebCore::DocumentMarkerController::forEach<DocumentMarkerController::IterationDirection::Forwards>):
(WebCore::DocumentMarkerController::forEach<DocumentMarkerController::IterationDirection::Backwards>):
(WebCore::addMarker):
* Source/WebCore/dom/DocumentMarkerController.h:
(WebCore::addMarker):
* Source/WebCore/page/writing-tools/WritingToolsController.h:
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::smartReplySessionDidReceiveTextWithReplacementRange):
(WebCore::WritingToolsController::compositionSessionDidReceiveTextWithReplacementRange):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView didBeginWritingToolsSession:contexts:]):
(-[WKWebView didEndWritingToolsSession:accepted:]):
(-[WKWebView compositionSession:didReceiveText:replacementRange:inContext:finished:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::compositionSessionDidReceiveTextWithReplacementRange):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/WebKitSwift.h:
* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceSmartReplyTextEffectCoordinator.h: Copied from Source/WebKit/WebKitSwift/WebKitSwift.h.
* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceSmartReplyTextEffectCoordinator.swift: Added.
(WKIntelligenceSmartReplyTextEffectCoordinator.onFlushCompletion):
(WKIntelligenceSmartReplyTextEffectCoordinator.hasActiveEffects):
(WKIntelligenceSmartReplyTextEffectCoordinator.startAnimation(for:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.requestReplacement(withProcessedRange:finished:characterDelta:operation:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.flushReplacements):
(WKIntelligenceSmartReplyTextEffectCoordinator.restoreSelection(_:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.hideEffects):
(WKIntelligenceSmartReplyTextEffectCoordinator.showEffects):
(WKIntelligenceSmartReplyTextEffectCoordinator.removeActiveEffects):
(WKIntelligenceSmartReplyTextEffectCoordinator.startReplacementAnimation(withProcessedRange:characterDelta:operation:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.reset):
(WKIntelligenceSmartReplyTextEffectCoordinator.updateTextChunkVisibility(_:visible:force:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.textPreview(for:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.updateTextChunkVisibility(_:visible:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.performReplacementAndGeneratePreview(for:effect:remainder:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.replacementEffectWillBegin(_:)):
(WKIntelligenceSmartReplyTextEffectCoordinator.replacementEffectDidComplete(_:)):
(async(_:)):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::compositionSessionDidReceiveTextWithReplacementRange):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/289354@main
